### PR TITLE
Generalize from IList to IEnumerable in MessageFactory, ChoiceFactory, and SuggestedActions.

### PR DESCRIPTION
**Pull request by JonathanFingold**
Update MessageFactory.SuggestedAction parameters to accept arbitrary enumerations.

### CODE REVIEW

File: libraries/Microsoft.Bot.Schema/SuggestedActions.cs [hunk 1]
```diff
@@ -35,10 +35,10 @@ public SuggestedActions()
         /// shown to.  These Ids are relative to the channelId and a subset of
         /// all recipients of the activity</param>
         /// <param name="actions">Actions that can be shown to the user</param>
-        public SuggestedActions(IList<string> to = default(IList<string>), IList<CardAction> actions = default(IList<CardAction>))
+        public SuggestedActions(IEnumerable<string> to = default(IEnumerable<string>), IEnumerable<CardAction> actions = default(IEnumerable<CardAction>))
         {
-            To = to;
-            Actions = actions;
+            To = to.ToList();
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Schema/SuggestedActions.cs]: These can be `null`, so the `.ToList()` call would fail with a `NullReferenceException`. I believe you should be able to remedy this easily with a null propagating operator:

```CSharp
To = to?.ToList();
```

Comment (drub0y) [reply libraries/Microsoft.Bot.Schema/SuggestedActions.cs]: I just noticed this actually broke one of the tests in the build, so we have some coverage of this scenario already at least.

Comment (JonathanFingold) [reply libraries/Microsoft.Bot.Schema/SuggestedActions.cs]: Good catch. Just updated the PR.

**Comment by JonathanFingold** [comment]
Just noticed this also applies to ChoiceFactory.ToChoices. Will add fix for that one presently.

File: libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs [hunk 2]
```diff
@@ -116,14 +116,14 @@ public static Activity Inline(List<Choice> choices, string text = null, string s
...
         }
 
-        public static Activity List(List<string> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
+        public static Activity List(IEnumerable<string> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
         {
             return List(ToChoices(choices), text, speak, options);
         }
 
-        public static Activity List(List<Choice> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
+        public static Activity List(IEnumerable<Choice> choiceList, string text = null, string speak = null, ChoiceFactoryOptions options = null)
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs]: Again, `choices` would seem like a better name than `choiceList` here.

File: libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs [hunk 3]
```diff
@@ -116,14 +116,14 @@ public static Activity Inline(List<Choice> choices, string text = null, string s
...
         {
-            choices = choices ?? new List<Choice>();
+            var choices = choiceList?.ToList() ?? new List<Choice>();
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs]: So, in the case of this method (there might be others), if we're accepting an `IEnumerable<T>` now... I would think it would behoove us to improve the implementation(s), when possible, such that we don't have to pay the price of the `.ToList()`.


For example, all we really need to do here is enumerate once over the incoming choices. So we can change this method to use a `foreach` instead with a separate `int index` variable that we increment manually each step and, in fact, you can only bother incrementing it inside of the `if(includeNumbers)` branch.

Also, while I'm here looking at this, this should really be changed to use a `StringBuilder` instead of `string` concats to reduce allocations/garbage.

File: libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs [hunk 4]
```diff
@@ -156,14 +156,14 @@ public static Activity List(List<Choice> choices, string text = null, string spe
...
         }
 
-        public static IMessageActivity SuggestedAction(List<string> choices, string text = null, string speak = null)
+        public static IMessageActivity SuggestedAction(IEnumerable<string> choices, string text = null, string speak = null)
         {
             return SuggestedAction(ToChoices(choices), text, speak);
         }
 
-        public static IMessageActivity SuggestedAction(List<Choice> choices, string text = null, string speak = null)
+        public static IMessageActivity SuggestedAction(IEnumerable<Choice> choiceList, string text = null, string speak = null)
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs]: Again, `choices` would seem like a better name than `choiceList` here.

File: libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs [hunk 5]
```diff
@@ -66,14 +66,14 @@ public static IMessageActivity ForChannel(string channelId, List<Choice> list, s
...
         }
 
-        public static Activity Inline(List<string> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
+        public static Activity Inline(IEnumerable<string> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
         {
             return Inline(ToChoices(choices), text, speak, options);
         }
 
-        public static Activity Inline(List<Choice> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
+        public static Activity Inline(IEnumerable<Choice> choiceList, string text = null, string speak = null, ChoiceFactoryOptions options = null)
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs]: `choiceList` doesn't seem as good a name as `choices` here, especially now that it's not an actual `List<T>`... did you mean to change???

Comment (JonathanFingold) [reply libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs]: I'll reverse the names so that the parameter keeps the name **choices** and the introduced local variable becomes **choiceList**, here and in the other places where I made the same change.

File: libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs (hunk 3 above)
Comment (JonathanFingold) [reply libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs]: I'll take a closer look. I was trying to avoid touching the implementation deeply, but I'll see what I can clean up here.

**Comment by JonathanFingold** [comment]
I've updated some of the implementation code. I don't have a favorite way of dealing with an enumerable and the Oxford comma, so I left that logic intact.

**Comment by JonathanFingold** [comment]
@cleemullins , Is there any reason to be using "\n" for new lines, as opposed to Environment.NewLine?

File: libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs [hunk 6]
```diff
@@ -85,88 +91,106 @@ public static Activity Inline(List<Choice> choices, string text = null, string s
...
 
-                txt += connector;
-                if (includeNumbers)
-                {
-                    txt += (index + 1).ToString() + ". ";
-                }
-                else
+            if (choices != null)
+            {
+                int count = choices.Where(c => c != null).Count();
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs]: See comment above about `.Count()` and double enumeration, but in this case there's truly no need to do the `.Count()` because `count` doesn't even appear to be used in this logic, right???

File: libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs [hunk 7]
```diff
@@ -85,88 +91,106 @@ public static Activity Inline(List<Choice> choices, string text = null, string s
...
-                if (index == (choices.Count - 2))
-                {
-                    connector = (index == 0 ? opt.InlineOr : opt.InlineOrMore) ?? string.Empty;
-                }
-                else
+            if (choices != null)
+            {
+                int count = choices.Where(c => c != null).Count();
+                int index = 1;
+                foreach (var choice in choices.Where(c => c != null))
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs]: I'm not sure how I feel about utilizing the `.Where(…)` to filter out potential null entries in the set. This feels to me like it should result in an exception and I would probably change the logic to something like:

```CSharp
if (choice == null)
{
    throw new ArgumentException("Set of choices contained a `null` instance which is not allowed.", nameof(choices));
}
```

File: libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs [hunk 8]
```diff
@@ -85,88 +91,106 @@ public static Activity Inline(List<Choice> choices, string text = null, string s
...
-                if (includeNumbers)
-                {
-                    txt += (index + 1).ToString() + ". ";
-                }
-                else
+            if (choices != null)
+            {
+                int count = choices.Where(c => c != null).Count();
+                int index = 1;
+                foreach (var choice in choices.Where(c => c != null))
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs]: See comment above about not using `.Where(…)` to hide potential `null` references and apply same here.

File: libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs [hunk 9]
```diff
@@ -85,88 +91,106 @@ public static Activity Inline(List<Choice> choices, string text = null, string s
...
                 {
-                    txt += "- ";
+                    string title = GetTitle(choice);
+                    sb.Append(separator);
+                    sb.Append((includeNumbers) ? $"{index}. " : "- ");
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs]: See comments above about avoiding mixing string interpolation with `StringBuilder`.

File: libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs [hunk 10]
```diff
@@ -85,88 +91,106 @@ public static Activity Inline(List<Choice> choices, string text = null, string s
...
+                    sb.Append(title);
+
+                    separator = Environment.NewLine + " ";
```
Comment (drub0y) [inline libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs]: This ends up generating a bunch of garbage `string`s (due to the concatenation) on each pass through the loop. Suggest moving the `Environment.NewLine + " "` into a static to avoid this _OR_ a second local variable _OR_, wrap this with a check that will only assign the separator to the new string on the first pass when it's `Length == 0` (`string.Empty` check).

Personally, for this pattern, what I've usually done in the past is _ALWAYS_ append the separator _AFTER_ the value and then outside the for loop you can back up the `StringBuffer` to truncate those characters off for the last value. You'd do that like this after the loop:

```CSharp
sb.Length -= separator.Length;
```

Make sense?

File: libraries/Microsoft.Bot.Builder.Core.Extensions/MessageFactory.cs [hunk 11]
```diff
@@ -145,16 +146,16 @@ public static IMessageActivity SuggestedActions(IList<string> actions, string te
...
+        public static IMessageActivity SuggestedActions(IEnumerable<CardAction> cardActions, string text = null, string ssml = null, string inputHint = null)
         {
             if (cardActions == null)
                 throw new ArgumentNullException(nameof(cardActions));
 
             IMessageActivity ma = Activity.CreateMessageActivity();
             SetTextAndSpeak(ma, text, ssml, inputHint);
 
-            ma.SuggestedActions = new SuggestedActions { Actions = cardActions };
+            ma.SuggestedActions = new SuggestedActions { Actions = cardActions?.ToList() };
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Core.Extensions/MessageFactory.cs]: The "?" doesn't seem needed. We already know cardActions is not null (due to the null check), to ToList() will always work.

File: libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs [hunk 12]
```diff
@@ -85,88 +91,106 @@ public static Activity Inline(List<Choice> choices, string text = null, string s
...
+            }
+            sb.Append(" ");
 
-                txt += connector;
-                if (includeNumbers)
-                {
-                    txt += (index + 1).ToString() + ". ";
-                }
-                else
+            if (choices != null)
```
[the remark this answers is not in the document]
Comment (cleemullins) [reply libraries/Microsoft.Bot.Builder.Prompts/Choices/ChoiceFactory.cs]: This whole package is about to go away as John Taylor mergest it into the Dialogs assembly.

File: libraries/Microsoft.Bot.Schema/SuggestedActions.cs [hunk 13]
```diff
@@ -35,10 +35,10 @@ public SuggestedActions()
         /// shown to.  These Ids are relative to the channelId and a subset of
         /// all recipients of the activity</param>
         /// <param name="actions">Actions that can be shown to the user</param>
-        public SuggestedActions(IList<string> to = default(IList<string>), IList<CardAction> actions = default(IList<CardAction>))
+        public SuggestedActions(IEnumerable<string> to = default(IEnumerable<string>), IEnumerable<CardAction> actions = default(IEnumerable<CardAction>))
         {
-            To = to;
-            Actions = actions;
+            To = to?.ToList();
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Schema/SuggestedActions.cs]: I don't understand why all of these null checks have been added. I don't understand why we're allowing Null's here at all, and I would *strongly* vote to remove that code.

File: libraries/Microsoft.Bot.Builder.Core.Extensions/MessageFactory.cs (hunk 11 above)
Comment (JonathanFingold) [reply libraries/Microsoft.Bot.Builder.Core.Extensions/MessageFactory.cs]: agreed

File: libraries/Microsoft.Bot.Schema/SuggestedActions.cs (hunk 13 above)
Comment (JonathanFingold) [reply libraries/Microsoft.Bot.Schema/SuggestedActions.cs]: fixing

Comment (JonathanFingold) [reply libraries/Microsoft.Bot.Schema/SuggestedActions.cs]: @cleemullins , Should the SuggestedActions construct throw if **to** is null or empty?

**Comment by JonathanFingold** [comment]
I only now realized that SuggestedActions is auto generated. So my changes to the file are irrelevant unless I want to create overloads in a SuggestedActionsEx file.

**Comment by JonathanFingold** [comment]
@drub0y , @cleemullins , Please take another look. I merged in John's changes and addressed all of your feedback.

File: libraries/Microsoft.Bot.Builder.Dialogs/Choices/ChoiceFactory.cs [hunk 14]
```diff
@@ -4,42 +4,50 @@
...
+            var list = new List<Choice>();
 
             // Find maximum title length
             var maxTitleLength = 0;
-            foreach (var choice in list)
+            foreach (var choice in choices)
             {
-                var l = choice.Action != null && string.IsNullOrEmpty(choice.Action.Title) ? choice.Action.Title.Length : choice.Value.Length;
-                if (l > maxTitleLength)
+                int len = GetTitle(choice).Length;
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Dialogs/Choices/ChoiceFactory.cs]: These changes seem to be a behaviorial. I don't see a revised test for this. 

I would suggest backing this change out of the PR, making this PR just abou the the IList -> IEnumerable change, and doing the behavioral changes in a different PR.

File: libraries/Microsoft.Bot.Builder.Dialogs/Choices/ChoiceFactory.cs [hunk 15]
```diff
@@ -85,88 +97,121 @@ public static Activity Inline(List<Choice> choices, string text = null, string s
...
+            var separator = string.Empty;
+            var sb = new StringBuilder();
+            if (text != null)
             {
-                var choice = choices[index];
-
-                var title = choice.Action != null && choice.Action.Title != null ? choice.Action.Title : choice.Value;
+                sb.Append(text);
+            }
+            sb.Append(' ');
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Dialogs/Choices/ChoiceFactory.cs]: More behavioral changes, that I think should be in a seperate PR.

File: libraries/Microsoft.Bot.Builder.Dialogs/Choices/ChoiceFactory.cs [hunk 16]
```diff
@@ -85,88 +97,121 @@ public static Activity Inline(List<Choice> choices, string text = null, string s
...
+                string title = GetTitle(choice);
+                sb.Append(separator);
+                if (opt.IncludeNumbers.Value)
                 {
-                    connector = opt.InlineSeparator ?? string.Empty;
+                    sb.Append('(');
+                    sb.Append(i + 1);
+                    sb.Append(") ");
                 }
+                sb.Append(title);
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Dialogs/Choices/ChoiceFactory.cs]: This whole block of new code is unexpected. Can we put this in a seperate, focused, PR that includes updated and/or relevant test coverage?

File: libraries/Microsoft.Bot.Builder.Dialogs/Choices/ChoiceFactory.cs [hunk 17]
```diff
@@ -85,88 +97,121 @@ public static Activity Inline(List<Choice> choices, string text = null, string s
...
 
-        public static Activity List(List<string> choices, string text = null, string speak = null, ChoiceFactoryOptions options = null)
+        /// <summary>
+        /// Gets a "normalized" title for a <see cref="Choice"/> object.
+        /// </summary>
+        /// <param name="choice">The choice object.</param>
+        /// <returns>The normalized title.</returns>
+        /// <exception cref="ArgumentNullException"/>
+        /// <exception cref="InvalidOperationException">The choice does not have a valid action title or value to use as a title.</exception>
+        private static string GetTitle(Choice choice)
```
Comment (cleemullins) [inline libraries/Microsoft.Bot.Builder.Dialogs/Choices/ChoiceFactory.cs]: I like the factoring of code here, the way this was extracted. 

... but let's try to keep each change seperate and tested.

File: tests/Microsoft.Bot.Builder.Prompts.Tests/ChoicePromptTests.cs [hunk 18]
```diff
@@ -0,0 +1,334 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.Prompts.Tests/ChoicePromptTests.cs]: This is a great new set of tests. I'm not clear why it's in a PR about changing IList to IEnumerable. 

Please put these tests in their own PR.

File: tests/Microsoft.Bot.Builder.Prompts.Tests/ConfirmPromptTests.cs [hunk 19]
```diff
@@ -0,0 +1,217 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (cleemullins) [inline tests/Microsoft.Bot.Builder.Prompts.Tests/ConfirmPromptTests.cs]: Same comment. Great tests. Let's make sure the functionality they are testing is in a clean PR.

**Comment by JonathanFingold** [comment]
Thanks @cleemullins . I'll close this PR and open a set of new, more focused ones. This grew out of a simple fix that...blossomed.

File: tests/Microsoft.Bot.Builder.Prompts.Tests/ChoicePromptTests.cs (hunk 18 above)
Comment (JonathanFingold) [reply tests/Microsoft.Bot.Builder.Prompts.Tests/ChoicePromptTests.cs]: I don't know why these are showing up as a change I'm making. These are from my pulling in recent changes from main. However, that will be moot soon.

**Comment by JonathanFingold** [comment]
@johnataylor , There is some potentially useful stuff here for the ChoiceFactory; however, allowing it to consume IEnumerable<Choice> may not be one of them. I'll leave it to you to ignore or possibly mine from the changes to ChoiceFactory that I'm discarding.
